main.c

```diff
@@ -89,15 +365,80 @@ int main(int argc, char** argv)
     memset(fcontent, 0, sizeof(fcontent));
     fread(fcontent + DOS_ADDR, fsize, 1, f);
 
-    err = uc_open(UC_ARCH_X86, UC_MODE_16, &uc);
-    if (err)
-    {
-        fprintf(stderr, "Cannot initialize unicorn\n");
+    typedef struct {
+      char signature[2]; /* == 0x5a4D */
+      unsigned short bytes_in_last_block;
+      unsigned short blocks_in_file;
+      unsigned short num_relocs;
+      unsigned short header_paragraphs;
+      unsigned short min_extra_paragraphs;
+      unsigned short max_extra_paragraphs;
+      unsigned short ss;
+      unsigned short sp;
+      unsigned short checksum;
+      unsigned short ip;
+      unsigned short cs;
+      unsigned short reloc_table_offset;
+      unsigned short overlay_number;
+    } EXE;
+
+    typedef struct {
+      unsigned short offset;
+      unsigned short segment;
+    } EXE_RELOC;
+
+
+       
+
+
+    err = uc_open (UC_ARCH_X86, UC_MODE_16, &uc);
+    if (err) {
+        fprintf (stderr, "Cannot initialize unicorn\n");
         return 1;
     }
 
+    EXE* exe = (EXE*)&fcontent[DOS_ADDR];
+    assert(exe->signature[0] == 'M');
+    assert(exe->signature[1] == 'Z');
+    unsigned int exe_data_start = exe->header_paragraphs * 16L;
+    unsigned int relocoff = DOS_ADDR / 0x10 + exe->header_paragraphs;
+    uint16_t t = exe->cs + relocoff;
+    uc_reg_write(uc, UC_X86_REG_CS, &t);
+    uint16_t ip = exe->ip;
+    t = exe->ss + relocoff;
+    uc_reg_write(uc, UC_X86_REG_SS, &t);
+    uc_reg_write(uc, UC_X86_REG_SP, &exe->sp);
+
+    printf("EXE start at 0x%04X:%04X (0x%04X:%04X)\n", exe->cs, exe->ip, exe->ss, exe->sp);
+
+    EXE_RELOC* relocs = (EXE_RELOC*)((uint8_t*)exe + exe->reloc_table_offset);
+    for(int i = 0; i < exe->num_relocs; i++) {
+      EXE_RELOC* reloc = &relocs[i];
+      uint16_t* v = (uint16_t*)&fcontent[relocoff * 16 + reloc->segment * 16 + reloc->offset];
+      printf("reloc: offset %X segment %X [0x%04X old]\n", reloc->offset, reloc->segment, *v);
+      *v += relocoff;
+      printf("reloc: offset %X segment %X [0x%04X new]\n", reloc->offset, reloc->segment, *v);
+    }
+
+
+#if 1
+// Some horrible patches to get around borland runtime
+
+// Specifically for "MTKFLASH by Joseph Lin, MTK 1998 (Ver 1.83c)"
+
+// Compare https://github.com/id-Software/wolf3d/blob/master/WOLFSRC/C0.ASM
+
+fcontent[0x182d1] = 0x90;
+fcontent[0x182d2] = 0x90;
+fcontent[0x182d3] = 0x90;
+
+fcontent[0x1007b+0] = 0x90;
+fcontent[0x1007b+1] = 0x90;
+fcontent[0x1007b+2] = 0x90;
+#endif
```

Comment:
This is the biggest issue at the moment - also prevents use of other tools like sdfndos.